### PR TITLE
Improve existing unit testing and add new ones

### PR DESCRIPTION
**Pull request by YGuerin**
GitHub Issue: #

## PR Type

What kind of change does this PR introduce?

- Bugfix
- Refactoring (no functional changes, no api changes)

## Description

Add unit testings to demonstrate new cases:
- Test if a method is called (Verify)
- Test the same method with different source of data (Theory)
- Use and setup mocking to avoid highly coupled code (Mock<>)
- Auto populate data (AutoFixture)

Bugfix:
- Avoid Search call if the search is not valid (SearchTerm.Length < 3 characters)

Functionnal update:
- Replace injection resolution with common pattern to support auto filled and mocked test data.

Naming update:
- Rename PostServiceTests to PostServiceShould to make the tests more "readable friendly"
  - The developer will now read **PostServiceShould > GetAllPosts** instead of **PostServiceTests > It_Should_GetAll** for example.

## PR Checklist 
Please check if your PR fulfills the following requirements:

- [ ] Interface members are XML documented
- [ ] Documentation (XML or comments) has been added and/or existing documentation has been updated
- [ ] [Architecture documents](./doc/Architecture.md) have been updated
- [ ] Tested on all relevant platforms

## Internal Issue (If applicable):

https://dev.azure.com/nventive/Practice%20committees/_workitems/edit/211221
https://dev.azure.com/nventive/Practice%20committees/_workitems/edit/211222

### CODE REVIEW

File: src/app/ApplicationTemplate.Shared/Presentation/Welcome/HomePageViewModel.cs [hunk 1]
```diff
@@ -63,7 +61,7 @@ public partial class HomePageViewModel : ViewModel
 
 		public IDynamicCommand NavigateToChuckNorrisSearchPage => this.GetCommandFromTask(async ct =>
 		{
-			await this.GetService<IStackNavigator>().Navigate(ct, () => new ChuckNorrisSearchPageViewModel());
+			await this.GetService<IStackNavigator>().Navigate(ct, () => new ChuckNorrisSearchPageViewModel(() => this.GetService<IChuckNorrisService>()));
```
Comment (MatFillion) [inline src/app/ApplicationTemplate.Shared/Presentation/Welcome/HomePageViewModel.cs:64]: Not sure we want to do this change, I think we explicitly decided each VM would resolve their own services.

@jeanplevesque any input?

File: src/app/ApplicationTemplate.Tests/Business/PostServiceShould.cs [hunk 2]
```diff
@@ -1,17 +1,15 @@
...
-using System.Text;
 using System.Threading.Tasks;
+using AutoFixture.Xunit2;
 using FluentAssertions;
 using Xunit;
 
 namespace ApplicationTemplate.Tests
 {
-	public class PostServiceTests : TestBase<IPostService>
+	public class PostServiceShould : TestBase<IPostService>
```
Comment (MatFillion) [inline src/app/ApplicationTemplate.Tests/Business/PostServiceShould.cs:9]: I like this change!

File: src/app/ApplicationTemplate.Tests/Business/PostServiceShould.cs [hunk 3]
```diff
@@ -20,22 +18,42 @@ public async Task It_Should_GetAll()
...
+			var postsNumber = posts.Count;
 
 			// Act
-			var result = await SUT.GetPost(DefaultCancellationToken, postId);
+			Func<Task<PostData>> act = () => SUT.GetPost(DefaultCancellationToken, givenId);
 
-			// Assert
-			result.Should().NotBeNull();
-			result.Id.Should().Be(postId);
+			if (givenId >= 1 && givenId <= postsNumber)
```
Comment (MatFillion) [inline src/app/ApplicationTemplate.Tests/Business/PostServiceShould.cs:37]: isnt it weird to have ifs in the test? does it mean it should be two tests instead?

File: src/app/ApplicationTemplate.Shared/Presentation/Welcome/HomePageViewModel.cs (hunk 1 above)
Comment (jeanplevesque) [reply src/app/ApplicationTemplate.Shared/Presentation/Welcome/HomePageViewModel.cs:64]: Yeah VMs have a ServiceProvider, so we shouldn't inject services.

Comment (YGuerin) [reply src/app/ApplicationTemplate.Shared/Presentation/Welcome/HomePageViewModel.cs:64]: I understand but I have no choice to do this (so far) to inject my mocked service to my test.
I'll try to find a way.

File: src/app/ApplicationTemplate.Tests/Business/PostServiceShould.cs (hunk 2 above)
Comment (YGuerin) [reply src/app/ApplicationTemplate.Tests/Business/PostServiceShould.cs:9]: Stolen from my Pluralsight lessons 👍 
I'll update on all other files in another pull request and add some documentation about the naming.

File: src/app/ApplicationTemplate.Shared/Presentation/Welcome/HomePageViewModel.cs (hunk 1 above)
Comment (jeanplevesque) [reply src/app/ApplicationTemplate.Shared/Presentation/Welcome/HomePageViewModel.cs:64]: If not, you can also override the `ConfigureHost` method on `TestBase`.

**Pull request closed without merging**